[Frontend] Added frontend with list product & create product

MartinLam-Oursky opened the pull request.

MartinLam-Oursky committed:
Added 'Add New Product' Page & Updated Navbar

MartinLam-Oursky committed:
Updated Form Submission Handler

MartinLam-Oursky committed:
Added Frontend

MartinLam-Oursky committed:
Added Shopping Cart

MartinLam-Oursky committed:
Added Scrolling for Shopping Cart

MartinLam-Oursky committed:
Finish add product page

MartinLam-Oursky committed:
Updated Home Page

1. Added Image for all products
2. Added Delete button for deleting product

MartinLam-Oursky committed:
Update navbar

MartinLam-Oursky committed:
Add container for home page

MartinLam-Oursky committed:
Add login page & route

MartinLam-Oursky committed:
Add register page

MartinLam-Oursky committed:
Add confirm dialog & icon

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,132 @@
+        });
+      }
+    )
+    var storageCart = localStorage.getItem("cart")
```
Use `const` here instead. In most cases `const` will suffice and offers better performance. 

In cases when you need a changeable variable, use `let`.

Some links explaining why we should always use `const` and `let`:
https://stackoverflow.com/questions/21237105/const-in-javascript-when-to-use-it-and-is-it-necessary
https://stackoverflow.com/questions/762011/whats-the-difference-between-using-let-and-var

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  addProductToCart(id: number) {
+    var arr = this.state.cart.length === 0 ? [] : [...this.state.cart]
```
`const`

elly0t commented on frontend/src/Components/NewProduct.tsx:
```diff
@@ -0,0 +1,92 @@
+
+    handleFormSubmit(event: React.FormEvent<HTMLFormElement>) {
+        event.preventDefault();
+        var form = document.querySelector("form");
```
`const`

elly0t commented on frontend/src/Components/NewProduct.tsx:
```diff
@@ -0,0 +1,92 @@
+        event.preventDefault();
+        var form = document.querySelector("form");
+        if( !form ) return false;
+        var formData = new FormData(form);
```
`const`

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,82 @@
+export default class Login extends React.Component{
+    handleFormSubmit(event: React.FormEvent<HTMLFormElement>) {
+        event.preventDefault();
+        var form = document.querySelector("form");
```
`const`

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,82 @@
+        event.preventDefault();
+        var form = document.querySelector("form");
+        if( !form ) return false;
+        var formData = new FormData(form);
```
`const`

elly0t commented on frontend/src/Components/Register.tsx:
```diff
@@ -0,0 +1,99 @@
+export default class Register extends React.Component{
+    handleFormSubmit(event: React.FormEvent<HTMLFormElement>) {
+        event.preventDefault();
+        var form = document.querySelector("form");
```
`const`

elly0t commented on frontend/src/Components/Register.tsx:
```diff
@@ -0,0 +1,99 @@
+        event.preventDefault();
+        var form = document.querySelector("form");
+        if( !form ) return false;
+        var formData = new FormData(form);
```
`const`

elly0t commented on frontend/src/Components/ShoppingCart.tsx:
```diff
@@ -0,0 +1,171 @@
+        if( index === -1 ) return
+        var copyOfArr = [...this.state.cart]
+        copyOfArr.splice(index, 1);
+        localStorage.setItem("cart", JSON.stringify(copyOfArr))
```
consider extract localStorage operations and put them into a centralised file, where their keys are enums/constants

elly0t replied:
ping

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,132 @@
+import "./App.css"
+
+export default class App extends React.Component {
+  state = ({
```
specify type for states

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,82 @@
+        if( !form ) return false;
+        var formData = new FormData(form);
+
+        fetch("http://127.0.0.1:8080/auth/login", {
```
similar to localStorage, extract them out and build a `ApiService` where the endpoints are centralised

elly0t commented on frontend/src/Components/Navbar.tsx:
```diff
@@ -0,0 +1,147 @@
+import MenuIcon from '@material-ui/icons/Menu'
+import { createStyles, makeStyles, Theme, withStyles } from '@material-ui/core/styles';
+
+const StyledMenu = withStyles({
```
wondering if this can be done through className and css files?

elly0t commented on frontend/src/Components/ShoppingCart.tsx:
```diff
@@ -0,0 +1,171 @@
+        localStorage.setItem("cart", "")
+        this.setState({
+            cart: [] as number[]
+        }, () => window.location.reload())
```
in React, updating the screen is normally achieved via changing props / states

elly0t commented on frontend/src/serviceWorker.ts:
```diff
@@ -0,0 +1,143 @@
+// This optional code is used to register a service worker.
```
think this file should be removed as we are not using it

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,132 @@
+      }).then(res => res.json())
+      console.log(result);
+      alert(result ? "Product Deleted." : "Failed to delete product.")
+      window.location.reload();
```
in React, updating the screen is normally achieved via changing props / states

MartinLam-Oursky committed:
Add new module for api call

MartinLam-Oursky committed:
delete serviceWorker

MartinLam-Oursky committed:
Change to use APIService for api call

MartinLam-Oursky committed:
Add Popup Box & rename style.css

MartinLam-Oursky committed:
Update login page to functional component

MartinLam-Oursky committed:
Update new product page to functional component

MartinLam-Oursky committed:
Update register page to functional component

MartinLam-Oursky committed:
update css file to seperate file

MartinLam-Oursky committed:
Update index page style

MartinLam-Oursky committed:
Update add new product page style

MartinLam-Oursky committed:
Move css files to new folder

MartinLam-Oursky committed:
Remove unused imports

MartinLam-Oursky committed:
Update login form style

MartinLam-Oursky committed:
Update register page style

MartinLam-Oursky committed:
Fix Popupbox class name

MartinLam-Oursky committed:
Update price field to accept float

MartinLam-Oursky committed:
Add Error & Loading Component

MartinLam-Oursky committed:
Fix bug & Update component

1. Fix popup box only able to show once bug
2. Update shopping cart functional component
3. Add PopupBoxProps interface
4. Add Product interface

MartinLam-Oursky committed:
Update navbar makestyle to css

MartinLam-Oursky committed:
Fix shopping cart stuck loading

MartinLam-Oursky committed:
Add component to manage cookies

MartinLam-Oursky committed:
Update Popupbox & other changes #25

1. Popup box can redirect after click the close button
2. Login & Register will save access token to cookies
3. Successful login will redirect to home page
4. Successful add product will redirect to home page
5. Successful register will redirect to home page
6. Update API Service to accept error

MartinLam-Oursky committed:
Update Buttons in navbar after login

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,70 @@
+const API_ENDPOINT = "http://localhost:8080"
+const API_PATH = {
+    AUTH: "/auth",
+    PRODUCT: "/product/"
```
better align `/product/` with `/auth`

elly0t commented on frontend/package.json:
```diff
@@ -0,0 +1,43 @@
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
+    "@material-ui/core": "^4.4.0",
```
Exact version would be safer, otherwise packages will be automatically upgraded to the most recent minor version upon any `npm install`. The caret tells npm to freeze the major version number but not the minor one.

elly0t commented on frontend/package.json:
```diff
@@ -0,0 +1,43 @@
+  "dependencies": {
+    "@material-ui/core": "^4.4.0",
+    "@material-ui/icons": "^4.2.1",
+    "@types/jest": "24.0.18",
```
the likes of `@types` should only exist in devDependencies

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,70 @@
+                headers: {
+                    'Accept': 'application/json',
+                    'Content-Type': 'application/json'
+                }
```
make duplicate fields of `headers` a const, say `COMMON_HEADERS`. Then when you need to add an extra field into it before firing the API, simply do:
```
headers: {
  ...COMMON_HEADERS,
  'custom field': `custom field value`
}
```

elly0t commented:
it appears that you haven't run the following command from oursky's create-react-app https://github.com/oursky/create-react-app-oursky when this frontend was set up
```
curl -sSL https://raw.githubusercontent.com/oursky/create-react-app-oursky/master/bootstrap.py | python3 - .
```

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,70 @@
+                method: "POST",
+                body: JSON.stringify(Object.fromEntries(data)),
+                headers: {
+                    'Accept': 'application/json',
```
First of all, oursky's current standard is to use double quotes.

It's best to align the styles - so when you have been using double quotes above (e.g. `"POST"`), change these single quotes to align.

Please ensure your prettier extension is working properly. Single quotes and other styling would become nonexistent if prettier had been set up properly in the first place.

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+	const [cart, setCart] = useState<number[]>([]);
+	const [popupProps, setPopupProps] = useState<PopupBoxProps | undefined>();
+
+	function updateStorageCart(){
```
Functions within an React functional components should be wrapped with a `useCallback` hook to avoid performance issue, as plain functions will be generated again upon every single render

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+	},[])
+
+	function addProductToCart(id: number) {
+		let arr = cart.length === 0 ? [] : [...cart]
```
when an array is declared with `const`, you can still change the value of its elements. You just can't resign the name of it.

examples:
```
const arr = []
arr.push(1)
arr.push(2)
console.log(arr) // prints [1, 2]
```
```
const arr = []
arr = [1, 2] // throws error!
```

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+		updateStorageCart()
+	},[])
+
+	function addProductToCart(id: number) {
```
not sure why only the product id is passed?

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+		arr.push(id)
+		setCart(arr)
+		localStorage.setItem("cart", JSON.stringify(arr))
+		const target = items.filter(v => v.id === id)
```
if the param of this function is the product object, the name of it can be directly accessed, instead of having to perform an O(n) for loop

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+	}
+
+	async function deleteProduct(id: number) {
+		if(!window.confirm("Are you sure you want to DELETE this product ?")){
```
reuse your self-made component pop up / modal?

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+				redirectTo: undefined
+			})
+
+		}catch(e){
```
again, please check if prettier is working properly

should be `} catch(e) {`

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+		}
+	}
+
+	return error ? <Error errorMessage = { error } /> : !isLoaded ? <Loading /> : (
```
instead of a nested if leading to 3 possibilities, perhaps the error should be shown within this screen?

elly0t commented on frontend/src/Components/App.tsx:
```diff
@@ -0,0 +1,195 @@
+					{
+						items.map((item: Product) => (
+							<Grid item key = {item.id}>
+								<Card className="myCard" >
```
all classNames should be implemented with css modules https://github.com/css-modules/css-modules

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,102 @@
+    const [loginResult, setLoginResult] = useState<string>("")
+    const [redirect, setRedirect] = useState<boolean>(false)
+
+    function handleFormSubmit (event: React.FormEvent<HTMLFormElement>){
```
`event.stopPropagation()` as well

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,102 @@
+        if( !form ) return false;
+
+        APIService.User.login(new FormData(form))
+        .then(result => {
```
the indentation is missing...

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,102 @@
+
+                <br />
+
+                <form onSubmit = {handleFormSubmit} >
```
form submission should be triggered with the Submit button below

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,102 @@
+            <CssBaseline />
+
+            {
+                loginResult !== "" && <PopupBox redirectTo = {undefined} title = "Login Result" message = {loginResult} open = {!!loginResult} onCloseClick = {() => setLoginResult("")} />
```
`onCloseClick={()  => setLoginResult("")}` a new function is created upon every render

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,102 @@
+    function handleFormSubmit (event: React.FormEvent<HTMLFormElement>){
+        event.preventDefault();
+        const form = document.querySelector("form");
+        if( !form ) return false;
```
return false is causing a warning `Not all code paths return a value.`

you don't have query and select the form if your submit function is triggered with the Submit button

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,102 @@
+                        label = "Username" 
+                    />
+
+                    <br />
```
I presume you are trying to create some vertical space here? margin-top or margin-bottom will suffice

elly0t commented on frontend/src/Components/Login.tsx:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    return redirect ? (
+            <div>
```
replace with React fragment https://reactjs.org/docs/fragments.html#short-syntax

more performant, as told by React Official doc 

> without adding extra nodes to the DOM

elly0t commented on frontend/src/Components/NewProduct.tsx:
```diff
@@ -0,0 +1,147 @@
+
+
+    function handleFormSubmit(event: React.FormEvent<HTMLFormElement>) {
+        event.preventDefault();
```
stopPropagation as well

elly0t commented:
- subfolders like `/components` and `interfaces` should be lower-cased
- should have 2 different folders `components` and `screens` differentiating them
- styles should be put together with components or screens
e.g. styles of `/screens/ScreenA.tsx` should be placed at a path `/screens/ScreenA.module.scss`
e.g. styles of `/component/ComponentB.tsx` should be placed at a path `/component/ComponentB.module.scss`

elly0t commented on frontend/src/Styles/Login.css:
```diff
@@ -0,0 +1,5 @@
+.MyPaper {
```
class name should be lower-case

elly0t commented on frontend/src/Styles/Error.css:
```diff
@@ -0,0 +1,6 @@
+.myPaper {
+    width: 100%;
```
let flex take care of the width and height

elly0t commented on frontend/src/Styles/ShoppingCart.css:
```diff
@@ -0,0 +1,5 @@
+.MyPaper {
+    padding: 10%;
```
pixels

elly0t commented on frontend/src/routes.tsx:
```diff
@@ -0,0 +1,23 @@
+export default (
+    <Router history = {history}>
+        <Navbar/>
+        <br />
```
margin will suffice

elly0t commented on frontend/src/Components/Register.tsx:
```diff
@@ -0,0 +1,113 @@
+            }else{
+                setSubmitResult("Successfully Registered.")
+                const date = new Date();
+                date.setTime(date.getTime() + (1000 * 60 * 10));
```
move to util functions

elly0t commented on frontend/src/Components/ShoppingCart.tsx:
```diff
@@ -0,0 +1,160 @@
+        setIsLoaded(true)
+    }, [updated])
+
+    function deleteProductFromCart(id: number) {
```
again, can pass the entire product in

elly0t commented on frontend/src/Components/ShoppingCart.tsx:
```diff
@@ -0,0 +1,160 @@
+        </Container>
+    ) : (
+            
+        <Container maxWidth="sm" component = "main">
```
apparently main is not a HTML element?

elly0t commented on frontend/src/Components/ShoppingCart.tsx:
```diff
@@ -0,0 +1,160 @@
+    let product: Product;
+    let subTotal = 0;
+
+    return !isLoaded ? (
```
1. the error can go into the Container below

2. create a reusable loading component, pass in the loading state and let it handle the logic

3. too many ternary operators will make the code less unreadable. Put the empty shopping cart and a non-empty one within the same container will reduce the number of IFs

elly0t commented on frontend/src/Components/ShoppingCart.tsx:
```diff
@@ -0,0 +1,160 @@
+                            subTotal += product.price;
+
+                            return (
+                                <div>
```
replace with React fragment <>

elly0t commented:
@MartinLam-Oursky to share styles, look for `@mixin` and `@include` in https://sass-lang.com/guide

css-module: https://github.com/css-modules/css-modules

MartinLam-Oursky committed:
Refactor project

1. Split components & screens
2. Each components & screens have its own subfolder
3. Update popupbox to use material-ui dialog
4. Add self-made confirm dialog
5. Add shopping cart interface for localStorage
6. Add eslint & prettier config
7. Update react-script to version 3.1.2(latest)
8. Update css to scss
9. Add Utils

MartinLam-Oursky committed:
Update to higher-order function

MartinLam-Oursky committed:
Update login to higher-order function

MartinLam-Oursky committed:
Remove unnecessary code

MartinLam-Oursky committed:
Delete register page unnecessary code

MartinLam-Oursky committed:
Update to use useCallback

MartinLam-Oursky committed:
Add Forgot Password frontend

MartinLam-Oursky committed:
Add Logout function

MartinLam-Oursky committed:
Update frontend for merchant & user

MartinLam-Oursky committed:
Update frontend for new backend design

1. Add merchant endpoint
2. Add merchant login & register
3. Fix bug

MartinLam-Oursky committed:
Add Manage Product page

MartinLam-Oursky committed:
Remove delete button in home screen

MartinLam-Oursky committed:
Update Add & Manage Product only visible to Merchant

MartinLam-Oursky committed:
Add check is merchant while mounting

MartinLam-Oursky committed:
Update to only show merchant's products

elly0t commented on frontend/package.json:
```diff
@@ -0,0 +1,53 @@
+  "dependencies": {
+    "@material-ui/core": "4.4.0",
+    "@material-ui/icons": "4.2.1",
+    "react": "^16.9.0",
```
better to given exact version to avoid auto version update

elly0t replied:
better to give exact version to avoid auto update

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,73 @@
+  "Content-Type": "application/json",
+};
+
+const AUTH_HEADER = (token: String) => {
```
`string`

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,73 @@
+  AUTH: "/auth",
+  PRODUCT: "/product/",
+};
+const HEADER = {
```
`HEADERS`

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,73 @@
+        method: "GET",
+        headers: {
+          ...HEADER,
+          ...AUTH_HEADER(token),
```
shouldn't this be returning a `string` instead of a `object`

elly0t commented on frontend/src/Interfaces/APIResponses.tsx:
```diff
@@ -0,0 +1,27 @@
+}
+
+export interface DeleteProductResult {
+  success: boolean;
```
success can be indicated simply by a status code 200

elly0t commented on frontend/src/Interfaces/APIResponses.tsx:
```diff
@@ -0,0 +1,27 @@
+  userId: number | undefined;
+  accessToken: string | undefined;
+  error: string | undefined;
+}
```
split into two responses instead

one for success - `LoginResponse`
one for error - `APIError`, which should be generic and applies to all APIs upon request failure
```js
export interface APIError {
  error: string
  details: string
}
```

elly0t replied:
apply such pattern to other responses as well

elly0t commented on frontend/src/Interfaces/DialogsProps.tsx:
```diff
@@ -0,0 +1,17 @@
+}
+
+export interface ConfirmDialogProps {
+  title: string | undefined;
```
shouldn't this be always a `string`?

elly0t commented on frontend/src/Interfaces/DialogsProps.tsx:
```diff
@@ -0,0 +1,17 @@
+import { MouseEventHandler } from "react";
+
+export interface AlertDialogProps {
+  title: string | undefined;
```
shouldn't this be always a `string`?

elly0t replied:
compulsory fields shouldn't be nullable

plus, nullable field should be indicated with:
```
export interface SomeProps {
  nullableField?: string
}
```

elly0t commented on frontend/src/Interfaces/DialogsProps.tsx:
```diff
@@ -0,0 +1,17 @@
+export interface AlertDialogProps {
+  title: string | undefined;
+  message: string | undefined;
+  open: boolean | undefined;
```
this is called `open` while the other is called `isOpen`, a bit inconsistent

elly0t commented on frontend/src/Interfaces/LoginData.tsx:
```diff
@@ -0,0 +1,4 @@
+export default interface LoginData {
+  email: String;
```
`string`

elly0t commented on frontend/src/Interfaces/PopupBoxProps.tsx:
```diff
@@ -0,0 +1,9 @@
+import { MouseEventHandler } from "react";
+
+export default interface PopupBoxProps {
+  title: string | undefined;
```
compulsory fields shouldn't be nullable

plus, nullable field should be indicated with:
```
export interface SomeProps {
  nullableField?: string
}
```

elly0t commented on frontend/src/Interfaces/RegisterData.tsx:
```diff
@@ -0,0 +1,5 @@
+export default interface RegisterData {
+  email: String;
```
`string`

elly0t commented on frontend/src/components/ConfirmDialog/index.tsx:
```diff
@@ -0,0 +1,42 @@
+}: ConfirmDialogProps) {
+  return (
+    <Dialog
+      open={!!isOpen}
```
this should always be a `boolean`, shouldn't be nullable

elly0t commented on frontend/src/routes.tsx:
```diff
@@ -0,0 +1,27 @@
+export default (
+  <Router history={history}>
+    <Navbar />
+    <br />
```
https://github.com/oursky/presentation-microservices-kotlin/pull/23#discussion_r327597451

elly0t replied:
we can use margin-bottom instead of a bunch of <br>s

MartinLam-Oursky committed:
Update to exact version

MartinLam-Oursky committed:
Update br to margin

MartinLam-Oursky committed:
Update variable name

MartinLam-Oursky committed:
Update variable name & type

MartinLam-Oursky committed:
Add update product function

MartinLam-Oursky committed:
Update home page layout

1. Add Image placeholder
2. Fix each product card size
3. Add change viewmode button

MartinLam-Oursky committed:
Update viewmode button style

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,86 @@
+      }).then(res => res.json()),
+
+    update: (id: number, data: NewProductData, token: string) => {
+      console.log(id);
```
get rid of these console.log

elly0t commented on frontend/src/screens/ShoppingCart/index.tsx:
```diff
@@ -0,0 +1,140 @@
+                      <IconButton
+                        edge="end"
+                        aria-label="delete"
+                        onClick={() =>
+                          deleteProductFromCart(product.id.toString())
+                        }
```
wrap this with `useCallback`

inline functions get re-generated upon every render

elly0t commented on frontend/src/components/AlertDialog/index.tsx:
```diff
@@ -0,0 +1,48 @@
+}: AlertDialogProps) {
+  return (
+    <Dialog
+      open={!!isOpen}
```
`isOpen` is already a boolean

elly0t commented on frontend/src/components/UpdateProductDialog/index.tsx:
```diff
@@ -0,0 +1,122 @@
+
+  return (
+    <Dialog
+      open={!!isOpen}
```
`isOpen` is already a boolean

elly0t commented on frontend/src/screens/ShoppingCart/index.tsx:
```diff
@@ -0,0 +1,140 @@
+                      {product.image && (
+                        <Avatar
+                          alt={product.name}
+                          src={`http://localhost:9000/images/${product.image}`}
```
store this with a variable which is parsed from `.env`

elly0t commented on frontend/src/APIService/index.tsx:
```diff
@@ -0,0 +1,86 @@
+import NewProductData from "../interfaces/NewProductData";
+import RegisterData from "../interfaces/RegisterData";
+import { LoginResult, RegisterResult } from "../interfaces/APIResponses";
+const API_ENDPOINT = "http://localhost:8080";
```
store this with a variable which is parsed from `.env`

elly0t commented on frontend/src/components/UpdateProductDialog/index.tsx:
```diff
@@ -0,0 +1,122 @@
+      <DialogContent>
+        {product.image && (
+          <CardMedia
+            image={"http://localhost:9000/images/" + product.image}
```
store this with a variable which is parsed from `.env`

elly0t commented on frontend/src/screens/Home/index.tsx:
```diff
@@ -0,0 +1,185 @@
+                  <CardMedia
+                    image={
+                      item.image
+                        ? "http://localhost:9000/images/" + item.image
```
store this with a variable which is parsed from `.env`

elly0t commented on frontend/src/screens/ManageProduct/index.tsx:
```diff
@@ -0,0 +1,320 @@
+                    <CardActionArea>
+                      {item.image && (
+                        <CardMedia
+                          image={"http://localhost:9000/images/" + item.image}
```
store this with a variable which is parsed from `.env`

MartinLam-Oursky committed:
Remove unnecessary operator

MartinLam-Oursky committed:
Add useCallback to functions

MartinLam-Oursky committed:
Update endpoint & image url to .env
